Add a command for running all pipeline stages on all files

glittershark opened the pull request:
Coming from Guard in ruby-land, one thing I often do is after working for a while adding some functionality to a module, I'll quickly run all the tests for the entire project - having tests run only for that file is nice while working as a quick feedback loop, but there's always the possibility that the change I've just introduced has broken some other, distant dependent part of the code, and I always like to run everything just as a quick last check before I open a PR. I'd been missing that functionality when working with Cortex, so I've gone ahead and implemented it here. I've done this by adding a separate callback to the `Cortex.Controller.Stage` behaviour, along with some additional branching in the `Cortex.Controller.run_pipeline` function to handle different kinds of things that might happen as part of a pipeline run (run_all vs run_file, maybe a couple more later?).

As you all know I'm still teaching myself Elixir/Erlang/OTP (in fact attempting to add this feature was part of that self-education endeavor), so it's possible (likely, even) that I've missed some idioms or obvious ways of doing things here and there - please let me know if that's the case and I'm obviously more than happy to address any feedback you might have.

It's also possible this isn't something you ever want as part of Cortex (it's doable from another terminal window via `mix test`, after all), which is also fine! Especially given that I didn't talk to anybody before I went off and wrote this, I wouldn't be offended if you decided to close it outright. I learned a lot and had a lot of fun writing it!

glittershark committed:
Refactor and break apart Cortex.Controller.run_pipeline

Refactor Cortex.Controller.run_pipeline to break its functionality apart
a little bit, to prepare for allowing different kinds of things to be
run as part of a pipeline

glittershark committed:
Add a command to reload and run tests for all modules

Add a method to the Cortex.Controller.Stage behaviour to run that stage
on all files, a cast case for the controller to call that method on all
stages, and a convenience function in the `Cortex` module to send that
cast to the controller server.

glittershark committed:
Add support in TestRunner for running all tests

Add support in Cortex.TestRunner for a call to run all tests

glittershark committed:
Add some light documentation for `Cortex.all`

OvermindDL1 commented:
I think it would be a useful option to run everything.  Perhaps as a configurable option.  Perhaps every X number of runs it runs all tests instead of just a limited set?

russmatney commented on lib/cortex/test_runner.ex:
```diff
@@ -76,35 +81,64 @@ defmodule Cortex.TestRunner do
-          |> Enum.reject(&(&1 == :ok))
+        case run_test_files([test_path]) do
+          :ok -> {:reply, :ok, state}
+          err = {:error, _} -> err
```
super nit-pick: we prefer pattern before var name for readability in these cases too - will add that to the styleguide

russmatney commented on lib/cortex/test_runner.ex:
```diff
@@ -76,35 +81,64 @@ defmodule Cortex.TestRunner do
+  def handle_call(:run_all, _from, state) do
+    case run_test_files() do
+      :ok -> {:reply, :ok, state}
+      err = {:error, _} -> err
```
same as above comment

russmatney commented on lib/cortex/test_runner.ex:
```diff
@@ -76,35 +81,64 @@ defmodule Cortex.TestRunner do
-              |> Enum.join("\n")
+  def run_test_files(_test_helper, []), do: :ok
+  def run_test_files(test_helper, files) do
+    files_to_load = [test_helper | files]
```
style nit: newlines breaking and following variable assignment here, should read:

```
files_to_load =
  [test_helper | files]

compiler_errors =
  # etc.
```

russmatney commented on lib/cortex/test_runner.ex:
```diff
@@ -76,35 +81,64 @@ defmodule Cortex.TestRunner do
+      |> Enum.reject(&(&1 == :ok))
+
+    with [] <- compiler_errors do
+      task = Task.async(ExUnit, :run, [])
```
same var assignment newline comment as above.

russmatney commented on lib/cortex/test_runner.ex:
```diff
@@ -76,35 +81,64 @@ defmodule Cortex.TestRunner do
 
-            {:reply, {:error, compiler_error_descs}, state}
-        end
+  defp all_test_files do
```
This works fine for typical elixir apps, but as written won't grab any test files when cortex is run in an umbrella root – you'll need to collect all the test files across all the umbrella/apps/**, or some similar method. I poked around for a bit but am not sure the cleanest way to do it. Note `Mix.Project.umbrella?` for checking on that case.

russmatney reviewed (commented):
Amazing job of simplifying the logic, the smaller functions are much better. Couple of nit-picks on style things, and it'd be nice to handle umbrella apps as well - I can take a harder look at doing that this week unless you beat me to it

russmatney committed:
feat(run_all): run_all gathering all test files for umbrella apps

For now, this uses Mix.Project.apps_paths() and expects all test files
to be in a "test" dir - it'd be nice to use each sub-app's
config[:test_paths], but we may never need it, as the rest of the test
files will have already been loaded anyway. Happy to dig further in when
that time comes. This also sets an infinite timeout on the genserver
call, as an umbrella's worth of tests won't always beat 5 seconds.
